session.go

```diff
@@ -808,14 +866,34 @@ func (s *session) handlePacketAfterClosed(p *receivedPacket) {
 	}
 }
 
+// MaybeRecover runs fail(errImmediate) if injection attack mitigation is turned off,
+// or fail(errTimeout) if no packets are received before AttackTimeout elapses
+func (s *session) maybeRecover(fail func(error), errImmediate error, errTimeout error) {
```

Comment:
The session struct has a "run loop" model and a central timer that is used to handle all wake-up things. That's more efficient, and avoids many many mutexes. Maybe you can fit this timeout into that structure too?